### PR TITLE
[SYCL-MLIR][NFC] Restructure Polygeist dialect

### PR DESCRIPTION
**Pull request by victor-eds**
Apply following changes to the dialect to follow MLIR dialects conventions:

- Change headers locations:
  - Dialect definition: "mlir/Dialect/Polygeist/IR/"
  - Transform passes: "mlir/Dialect/Polygeist/Transforms/"
  - Conversion passes: "mlir/Conversion/PolygeistTo<dialect>/"
  - Dialect utils: "mlir/Dialect/Polygeist/Utils/<file>.h"
- Create a library for each conversion pass.
- Conversion passes are defined under the `mlir` namespace.

This commit does not intend to be a complete refactoring of the dialect. This is a work in progress, not intending to refactor any code.

### CODE REVIEW

**Review by whitneywhtsang — APPROVED** [review]

**Comment by victor-eds** [comment]
Failing tests are not related (missing `-lze`)

File: polygeist/include/mlir/Conversion/PolygeistPasses.td [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#ifndef MLIR_CONVERSION_POLYGEISTPASSES
+#define MLIR_CONVERSION_POLYGEISTPASSES
+
+include "mlir/Pass/PassBase.td"
+
+def ConvertPolygeistToLLVM : Pass<"convert-polygeist-to-llvm", "mlir::ModuleOp"> {
+  let summary = "Convert scalar and vector operations from the Standard to the "
```
Comment (etiotto) [inline polygeist/include/mlir/Conversion/PolygeistPasses.td:7]: Is the comment accurate ?  I would expect a pass named `ConvertPolygeistToLLVM` to convert polygeist operations to LLVM. Other dialect operations should be converted by their respective conversion passes.

File: polygeist/include/mlir/Conversion/PolygeistPasses.td [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                "LLVM dialect";
+  let description = [{
+    Convert standard operations into the LLVM IR dialect operations.
```
Comment (etiotto) [inline polygeist/include/mlir/Conversion/PolygeistPasses.td:10]: ditto as above

File: polygeist/include/mlir/Dialect/Polygeist/IR/PolygeistBase.td [hunk 3]
```diff
@@ -0,0 +1,20 @@
+//=== PolygeistBase.td - Base definitions for polygeist dialect -*- tablegen =//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/IR/PolygeistBase.td:7]: We need the LLVM standard copyright

File: polygeist/include/mlir/Dialect/Polygeist/IR/PolygeistOps.td [hunk 4]
```diff
@@ -1,4 +1,4 @@
-//===- Polygeist.td - Polygeist dialect ops ----------------*- tablegen -*-===//
+//===- PolygeistOps.td - Polygeist op definitions ----------*- tablegen -*-===//
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/IR/PolygeistOps.td:1]: We need the LLVM standard copyright

File: polygeist/include/mlir/Conversion/PolygeistPasses.td [hunk 5]
```diff
@@ -0,0 +1,53 @@
+#ifndef MLIR_CONVERSION_POLYGEISTPASSES
```
Comment (etiotto) [inline polygeist/include/mlir/Conversion/PolygeistPasses.td:1]: Need to add a description plus we need the LLVM standard copyright

File: polygeist/include/mlir/Dialect/Polygeist/Transforms/Passes.h [hunk 6]
```diff
@@ -1,14 +1,28 @@
-#ifndef POLYGEIST_DIALECT_POLYGEIST_PASSES_H
-#define POLYGEIST_DIALECT_POLYGEIST_PASSES_H
+#ifndef MLIR_DIALECT_POLYGEIST_TRANSFORMS_PASSES_H
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Transforms/Passes.h:1]: Pls add header and copyright statement.

File: polygeist/include/mlir/Dialect/Polygeist/Utils/BarrierUtils.h [hunk 7]
```diff
@@ -1,24 +1,25 @@
-//===- BarrierUtil.h - Utilities for barrier removal --------* C++ -*-===//
+//===- BarrierUtils.h - Utilities for barrier removal ------------* C++ -*-===//
 //
 // Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Utils/BarrierUtils.h:8]: add copyright to all files that do not have it pls.

File: polygeist/lib/Dialect/Polygeist/Transforms/BareMemRefToLLVM.cpp [hunk 8]
```diff
@@ -0,0 +1,337 @@
+//===- BareMemRefToLLVM.cpp - MemRef to LLVM with bare ptr call conv ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/BareMemRefToLLVM.cpp:8]: add copyright

File: polygeist/include/mlir/Conversion/PolygeistPasses.td (hunk 1 above)
Comment (victor-eds) [reply polygeist/include/mlir/Conversion/PolygeistPasses.td:7]: Fixed doc.

File: polygeist/include/mlir/Conversion/PolygeistPasses.td (hunk 2 above)
Comment (victor-eds) [reply polygeist/include/mlir/Conversion/PolygeistPasses.td]: Deleted long description. Most dialects don't have a long description for such passes. We can explain we need bare-ptr conversion when we enforce it.

File: polygeist/include/mlir/Dialect/Polygeist/IR/PolygeistBase.td (hunk 3 above)
Comment (victor-eds) [reply polygeist/include/mlir/Dialect/Polygeist/IR/PolygeistBase.td:7]: Added to all files.

**Comment by victor-eds** [comment]
Test failing with:
```
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  No device of requested type available. -1 (PI_ERROR_DEVICE_NOT_FOUND)
```

**Review by etiotto — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
Rebased to latest sycl-mlir commit to refresh testing.

**Pull request merged**